src/ukrdc_cupid/core/store/models/structure.py

```diff
@@ -0,0 +1,289 @@
+
+        Args:
+            child_node (Type[Node]): _description_
+            xml_attr (str): _description_
+            session (Session): _description_
+            sequential (bool, optional): _description_. Defaults to False.
+        """
+
+        # Step into the xml_file and extract the xml containing incoming data
+        xml_items = self.xml
+        for attr in xml_attr.split("."):
+            if isinstance(xml_items, list):
+                # this is nessary because of the weirdness of xsdata
+                xml_items = xml_items[0]
+
+            if xml_items:
+                xml_items = getattr(xml_items, attr, None)
+
+        mapped_ids = []
+        if xml_items:
+            if not isinstance(xml_items, list):
+                # for convenience treat singular items as a list
+                xml_items = [xml_items]
+
+            for seq_no, xml_item in enumerate(xml_items):
+                # Some item are sent in sequential order this order implicitly sets the keys
+                # there is a possibility here to sort the items before enumerating them
+                node_object = child_node(xml=xml_item)  # type:ignore
+
+                # map to existing object or create new
+                id = node_object.map_to_database(session, self.pid, seq_no)
+                mapped_ids.append(id)
+
+                # add parent info
+                # add any foreign keys, enumerations or data which doesn't come
+                # from the xml.
+                parent_data = self.generate_parent_data(seq_no)
+                for attr, value in parent_data.items():
+                    if hasattr(node_object.orm_object, attr):
+                        setattr(node_object.orm_object, attr, value)
+
+                # update new or existing using the
+                node_object.map_xml_to_orm(session)
+
+                # modify update_date and updated status
+                node_object.updated_status()
+
+                # append to parent
+                self.mapped_classes.append(node_object)
+
+        # if there are already objects mapped to record harmonise by deleting anything not in incoming
+        # xml file. This should be skipped for singular and manditory items (like patient) by setting
+        # self.sqla_relationship to None.
+        sqla_relationship = child_node.sqla_mapped()  # type : ignore
+        if sqla_relationship:
+            self.add_deleted(sqla_relationship, mapped_ids)
+
+    def add_deleted(self, sqla_mapped: str, mapped_ids: List[str]) -> None:
+        # stage records for deletion using the lazy mapping and a list of ids of records in the file
+        # this is only needed if the is a one to many relationship between parent and child
+        # this function stages for deletion objects which appear mapped but don't appear in incoming file
+        # This highlights a problem with the idx method creating keys. What if an item in the middle of the
```

Comment:
A PK based on certain attributes would be 'better' but I can only think there was some reason we didn't do that. I think this method works, just that it causes more 'churn' that is strictly neccesary.

Say we have

1. 01/01/2000 Haemodialysis
2. 01/02/2000 PD
3. 01/03/2000 Transplant
4. 01/04/2000 Conservative Management

Then the next file is -

1. 01/01/2000 Haemodialysis
2. 01/02/2000 PD
3. 01/04/2000 Conservative Management

Then I would expect the actions on import to be -

1. No Change
2. No Change
3. Update Row to replace details with that for the Conservative Management record
4. Delete Row

So data wise we're fine. There is a problem that any trigger based creation/update dates we're using won't reflect the date the records changed in the source system but I think we just have to acknowledge their limitations.

Reply:
Would this require changing all the keys in the existing database? That sounds like it might be a bit of a time commitment?

Reply:
Changing to a Composite PK method would.

I think ultimately we should do this as there are weaknesses in the current approach such as that we don't have triggers to update the ID if the fields themselves are updated.

For now just keep things as-is?